subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_security.c

```diff
@@ -181,16 +223,15 @@ static int load_credentials_to_modem(struct lwm2m_ctx *ctx)
 	}
 
 out:
-	if (fmode != LTE_LC_FUNC_MODE_POWER_OFF || fmode != LTE_LC_FUNC_MODE_OFFLINE) {
-		LOG_INF("Restoring modem connection");
-		/* I need to use the blocking call, because in next step
-		 * LwM2M engine would create socket and call connect()
-		 */
-		int err = lte_lc_connect();
-
-		if (err) {
-			LOG_ERR("lte_lc_connect() failed %d", err);
+	LOG_INF("Switching LTE and GNSS on once it's allowed...");
+	while (1) {
```

Comment:
I don't think depending on the return value of the callback will work for applications that doesn't call the link controller APIs in the callback itself.

The Link controller exposes a macro that allows a callback to be registered upon a change in functional mode.
`LTE_LC_ON_CFUN`. The you could simply depend on a semaphore in that callback (with a timeout) that gives the semaphore.

Roughly something like this:

```
LTE_LC_ON_CFUN(cfun_monitor, on_cfun, NULL);

static void on_cfun(enum lte_lc_func_mode mode, void *ctx)
{
	if (mode == LTE_LC_FUNC_MODE_NORMAL) {
		k_sem_give(&cfun_normal_sem);
	} else if (mode == LTE_LC_FUNC_MODE_OFFLINE || LTE_LC_FUNC_MODE_POWER_OFF) {
		k_sem_give(&cfun_offline_sem);
	}
}

/* Inside load_credentials_to_modem */

if (lte_lc_func_mode_get(&fmode)) {
	LOG_ERR("Failed to read modem functional mode");
	return -EFAULT;
}

if (fmode != LTE_LC_FUNC_MODE_POWER_OFF || fmode != LTE_LC_FUNC_MODE_OFFLINE) {
	/* Notify application that modem needs to be put in offline. */
	/* Clear semaphores. */
	k_sem_reset(&cfun_normal_sem);
	k_sem_reset(&cfun_offline_sem);
	mm.cb()

	/* Wait for modem to be put into offline mode. Semphore will be given in on_cfun(). */
	if (k_sem_take(&cfun_offline_sem, K_SECONDS(CONFIG_TIME_TO_WAIT_FOR_OFFLINE)) != 0) {
		/* If semaphore is not given within the timeout, an error is propagated. */
		return -ETIMEDOUT;
	}
}

/* Write credentials */
/* Credential writing finished. */
if (lte_lc_func_mode_get(&fmode)) {
	LOG_ERR("Failed to read modem functional mode");
	return -EFAULT;
}

if (fmode != LTE_LC_FUNC_MODE_NORMAL) {
	/* Notify application that modem needs to be put in offline. */
	/* Clear semaphores. */
	k_sem_reset(&cfun_normal_sem);
	k_sem_reset(&cfun_offline_sem);
	mm.cb()

	/* Wait for modem to be put into offline mode. Semphore will be given in on_cfun(). */
	if (k_sem_take(&cfun_normal_sem, K_SECONDS(CONFIG_TIME_TO_WAIT_FOR_OFFLINE)) != 0) {
		/* If semaphore is not given within the timeout, an error is propagated. */
		return -ETIMEDOUT;
	}
}
```

Reply:
Ok, I see. This works as well, it will require some additional logic in the app but it shouldn't be too much. Thanks for clarifying.